[Cookie Store API] Ensure event listener uses cookie creation time to decide whether it was deleted or changed

RupinMittal opened the pull request:
#### 90ca349aa08cc917dcf128f61cd9cc061197a315
<pre>
[Cookie Store API] Ensure event listener uses cookie creation time to decide whether it was deleted or changed
<a href="https://bugs.webkit.org/show_bug.cgi?id=285628">https://bugs.webkit.org/show_bug.cgi?id=285628</a>
<a href="https://rdar.apple.com/142518999">rdar://142518999</a>

Reviewed by Sihui Liu.

When the event listener is fired, we must decide whether the cookie
was changed or deleted so we can fire the correct type of event.

Current logic (incorrect):
if (cookie.expires &lt; currentTime) --&gt; it was deleted
else --&gt; it was changed

This is wrong because currentTime is the time of the event firing, not cookie creation.
We see this causing a problem in the test site. If the expiry time is really small,
say 0.1 seconds from now, the event will fire after the expiry, so we&apos;ll say the cookie
was deleted instead of changed. This is flaky as well. We should instead be comparing
with the creation time of the cookie.

New logic (correct):
if (cookie.expires &lt;= cookie.created) --&gt; it was deleted
else --&gt; it was changed

We need to expire in the (cookie.expires = cookie.created) case because of
the case where a cookie is set using document.cookie with max-age=0. In that
case, the cookie is expires as soon as it&apos;s set, which is considered deletion
(see the wpt test cookie-store/change_eventhandler_for_document_cookie.https.window.html)

This is verified by testing using the test case in the radar.

* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::cookiesAdded):

Canonical link: <a href="https://commits.webkit.org/288908@main">https://commits.webkit.org/288908@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/507c6e37bad56ee6281623fe67f2113a1e001bfc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84762 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4487 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39150 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/89901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/35814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86847 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4576 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12462 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/89901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/35814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87807 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/3468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/77024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/89901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/3347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/31243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34888 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/32049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/91277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/12101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/91277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/12328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/72835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/91277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/16379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/3549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13207 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/12053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/11887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/15381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/13633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

szewai reviewed (approved):
Is it possible to add a test?

RupinMittal committed:
[Cookie Store API] Ensure event listener uses cookie creation time to decide whether it was deleted or changed
https://bugs.webkit.org/show_bug.cgi?id=285628
rdar://142518999

Reviewed by Sihui Liu.

When the event listener is fired, we must decide whether the cookie
was changed or deleted so we can fire the correct type of event.

Current logic (incorrect):
if (cookie.expires < currentTime) --> it was deleted
else --> it was changed

This is wrong because currentTime is the time of the event firing, not cookie creation.
We see this causing a problem in the test site. If the expiry time is really small,
say 0.1 seconds from now, the event will fire after the expiry, so we'll say the cookie
was deleted instead of changed. This is flaky as well. We should instead be comparing
with the creation time of the cookie.

New logic (correct):
if (cookie.expires <= cookie.created) --> it was deleted
else --> it was changed

We need to expire in the (cookie.expires = cookie.created) case because of
the case where a cookie is set using document.cookie with max-age=0. In that
case, the cookie is expires as soon as it's set, which is considered deletion
(see the wpt test cookie-store/change_eventhandler_for_document_cookie.https.window.html)

This is verified by testing using the test case in the radar.

* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::cookiesAdded):

Canonical link: https://commits.webkit.org/288908@main
